apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/FDADrugUtils.java

```diff
@@ -0,0 +1,152 @@
+      drugCodeLookupMissingFailures.add(claimDrugCode);
+      LOGGER.info(
+          "No national drug code value (PRODUCTNDC column) match found for drug code {} in"
+              + " resource {}.",
+          claimDrugCode,
+          "fda_products_utf8.tsv");
+    }
+
+    return null;
+  }
+
+  /**
+   * Reads all the <code>PRODUCTNDC</code> and <code>SUBSTANCENAME</code> fields from the FDA NDC
+   * Products file which was downloaded during the build process.
+   *
+   * <p>See {@link gov.cms.bfd.server.war.FDADrugDataUtilityApp} for details.
+   *
+   * @return a map with drug codes and fields
+   */
+  public Map<String, String> readFDADrugCodeFile() {
+    Map<String, String> ndcProductHashMap = new HashMap<String, String>();
+    try (final InputStream ndcProductStream =
+            Thread.currentThread()
+                .getContextClassLoader()
+                .getResourceAsStream(FDADrugDataUtilityApp.FDA_PRODUCTS_RESOURCE);
+        final BufferedReader ndcProductsIn =
+            new BufferedReader(new InputStreamReader(ndcProductStream))) {
+      /*
+       * We want to extract the PRODUCTNDC and PROPRIETARYNAME/SUBSTANCENAME from the FDA Products
+       * file (fda_products_utf8.tsv is in /target/classes directory) and put in a Map for easy
+       * retrieval to get the display value which is a combination of PROPRIETARYNAME &
+       * SUBSTANCENAME
+       */
+      String line = "";
+      ndcProductsIn.readLine();
+      while ((line = ndcProductsIn.readLine()) != null) {
+        String ndcProductColumns[] = line.split("\t");
+        try {
+          String nationalDrugCodeManufacturer =
+              StringUtils.leftPad(
+                  ndcProductColumns[1].substring(0, ndcProductColumns[1].indexOf("-")), 5, '0');
+          String nationalDrugCodeIngredient =
+              StringUtils.leftPad(
+                  ndcProductColumns[1].substring(
+                      ndcProductColumns[1].indexOf("-") + 1, ndcProductColumns[1].length()),
+                  4,
+                  '0');
+          // ndcProductColumns[3] - Proprietary Name
+          // ndcProductColumns[13] - Substance Name
+          ndcProductHashMap.put(
+              String.format("%s-%s", nationalDrugCodeManufacturer, nationalDrugCodeIngredient),
+              ndcProductColumns[3] + " - " + ndcProductColumns[13]);
+        } catch (StringIndexOutOfBoundsException e) {
+          continue;
+        }
+
+        if (includeFakeDrugCode) {
+          appendFDATestCode(ndcProductHashMap);
+        }
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException("Unable to read NDC code data.", e);
+    }
+    return ndcProductHashMap;
+  }
+
+  private void appendFDATestCode(Map<String, String> ndcProductHashMap) {
+    String line =
+        "0000-0000_000000zz-0zz0-0z00-zzz0-0z00zzz00000\t0000-0000\tFAKE DRUG\tFake Diluent\t\tfake\tFAKE SOLUTION\tFake\t0\t\tFAK\tFAK000000\tFake Company\tWATER\t1\tmL/mL\t\t\tN\t00000000";
```

Comment:
if we end up just using part of this string do we need this? can we just put the right fake things into the map and be done?

Reply:
done